### DISCUSSION TITLE
example of using _createMoorIsolate

### DISCUSSION THREAD

**Discussion by matthewkooshad** [simolus3/drift#1257] [Q&A] [answered]
from https://moor.simonbinder.eu/docs/advanced-features/isolates/,
i'm not using VmDatabase.memory(), but rather an on-disk db, since i need the data to persist when the flutter app is closed.

from https://moor.simonbinder.eu/docs/advanced-features/isolates/#initialization-on-the-main-thread, 
i see Future<MoorIsolate> _createMoorIsolate() async , but i don't see an example of calling this. could i see an example of calling this?

in my main method, i started to put the line below there to set a global Database db variable for me to use this isolate db, but i need to see how to use this with _createMoorIsolate.
Future<void> main() async {
  db = Database.connect(DatabaseConnection.delayed(_connectAsync()));

as i noticed _connectAsync from that doc returns a Future<DatabaseConnection>
but, _createMoorIsolate from that doc returns a Future<MoorIsolate>

**Comment by simolus3** [comment, marked as answer]
`DatabaseConnection.delayed` is the way to go. It takes a future as an argument and returns a connection that is ready once the future completes.
So you should be able to just setup the database like this:

```dart
void main() {
  db = Database.connect(DatabaseConnection.delayed(_createMoorIsolate()));
  // ...
}
```

> as i noticed _connectAsync from that doc returns a Future
> but, _createMoorIsolate from that doc returns a Future

I don't see a contradiction here :D If you want to turn a `Future<DatabaseConnection>` into a `DatabaseConnection`, use `DatabaseConnection.delayed`.
If that doesn't work for you please let me know.

**Reply by matthewkooshad** [reply]
for the _createMoorIsolate() param, i get this error:
The argument type 'Future<MoorIsolate>' can't be assigned to the parameter type 'FutureOr<DatabaseConnection>'.dart(argument_type_not_assignable)

my Database class does have this ctor:
Database.connect(DatabaseConnection connection) : super.connect(connection);

and the _createMoorIsolate method i copied verbatim from the docs link in this thread.

**Reply by simolus3** [reply]
The constructor and the `_createMoorIsolate` method both look correct to me.

> The argument type 'Future' can't be assigned to the parameter type 'FutureOr'.dart(argument_type_not_assignable)

Uh, what? `Future` definitely extends `FutureOr`. I really don't understand how this could happen. Do you get the same error if you launch your app like that (even if the analyzer complains about that usage)?. Do you get any errors in the definition of `_createMoorIsolate`?

**Reply by matthewkooshad** [reply]
sorry, the error now shows for that param as:
The argument type 'Future<MoorIsolate>' can't be assigned to the parameter type 'FutureOr<DatabaseConnection>'.dart(argument_type_not_assignable)

no errors showing for _createMoorIsolate , _startBackground, or _IsolateStartRequest blocks of code.

i'll try to build and launch in a moment and let you know here if it does work.

**Reply by matthewkooshad** [reply]
sorry, in my OP, the part after Future was removed by this text editor...:
as i noticed _connectAsync from that doc returns a `Future<DatabaseConnection>`
but, _createMoorIsolate from that doc returns a `Future<MoorIsolate>`

**Reply by matthewkooshad** [reply]
Launching lib\main.dart on Android SDK built for x86 in debug mode...
lib\main.dart:1
lib/db/database.dart:29:52: Error: The argument type `'Future<MoorIsolate>'` can't be assigned to the parameter type `'FutureOr<DatabaseConnection>'`.
 - 'Future' is from 'dart:async'.
 - 'MoorIsolate' is from 'package:moor/isolate.dart' ('../../flutter/.pub-cache/hosted/pub.dartlang.org/moor-4.3.2/lib/isolate.dart').
 - 'DatabaseConnection' is from 'package:moor/src/runtime/api/runtime_api.dart' ('../../flutter/.pub-cache/hosted/pub.dartlang.org/moor-4.3.2/lib/src/runtime/api/runtime_api.dart').
  db = Database.connect(DatabaseConnection.delayed(_createMoorIsolate()));

FAILURE: Build failed with an exception.`

**Reply by simolus3** [reply]
Of course, sorry. I should have caught this earlier. You can use `.connect()` on the `MoorIsolate` to connect to it:

```dart
DatabaseConnection _createMoorIsolateAndConnect() {
  return DatabaseConnection.delayed(() async {
    final isolate = await _createMoorIsolate();
    return await isolate.connect();
  }());
}
```

**Reply by matthewkooshad** [reply]
![image](https://user-images.githubusercontent.com/6706761/121529426-00d92000-c9c2-11eb-9b41-3d69d7dcfe38.png)

**Reply by matthewkooshad** [reply]
i get that error every time now that i try to use the app. i guess it's with the _createMoorIsolate from the docs that uses getApplicationDocumentsDirectory [from path_provider]

**Reply by simolus3** [reply]
Tiny steps... In the debugger, you see that `getApplicationDocumentsDirectory()` is called from the main isolate, right? Do the path provider operations fail in general for you? Maybe that's a bug in that plugin.

**Reply by matthewkooshad** [reply]
```dart
Future<void> main() async {
  initDatabase();

  await initServices();

  runApp(const MyApp());
}
```

i saw that main moves forward with initServices while my initDatabase didn't finish, as i set some breakpoints in initDatabase code. also, in the call stack, i saw separate entries, though not very clear how to see that it's a separate thread there.

a visual of initDatabase:
```dart
void initDatabase() {
  db = Database.connect(DatabaseConnection.delayed(_createMoorIsolateAndConnect()));
}

DatabaseConnection _createMoorIsolateAndConnect() {
  return DatabaseConnection.delayed(() async {
    final isolate = await _createMoorIsolate();
    return isolate.connect();
  }());
}
```

in your code in the doc, you have `final dir = await getApplicationDocumentsDirectory();` , were you able to get this to work? i have never noticed an issue with path provider plugin, and it worked just fine for me in the other ctor that i have:
```dart
class Database extends _$Database {
  Database._() : super(_openConnection()); //is this other ctor conflicting with the one that my initDatabase is using?

  Database.connect(DatabaseConnection connection) : super.connect(connection); 
...
}

LazyDatabase _openConnection() {
  return LazyDatabase(() async {
    final dbFolder = await getApplicationDocumentsDirectory();
    final file = File(p.join(dbFolder.path, 'db.sqlite'));
    return VmDatabase(file, logStatements: logStatements);
  });
}
```

**Reply by simolus3** [reply]
Ah sorry, I think the problem is that you're creating the database before the messenger has been initialized. Try calling `WidgetsFlutterBinding.ensureInitialized` at the start of your `main` method.

Btw, you can make `initDatabase` async and use

```dart
Future<void> initDatabase() async {
  final isolate = await _createMoorIsolate();
  final connection = await isolate.connect();
  db = Database.connect(connection);
}   
```

Then, just await `initDatabase` in `main` since you're in an async context already.

**Reply by matthewkooshad** [reply]
thanks, after doing those 2 things, i now see on the call stack both: main and _startBackground simultaneously running!

also, my existing db methods appear working the same as before.

**Reply by matthewkooshad** [reply]
i also checked by removing `WidgetsFlutterBinding.ensureInitialized` from the start of my `main` method, and the problem with getApplicationDocumentsPath shows again just as before.

i read a bit here about this snippet:
https://stackoverflow.com/questions/63873338/what-does-widgetsflutterbinding-ensureinitialized-do/63873689
but, it's not clear to me, if this is a _hack(?)_ that would be fine for production version or if i should be looking into another way to getApplicationDocumentsPath.. -- thoughts? thanks

**Reply by simolus3** [reply]
`getApplicationDocumentsPath` works by sending a [platform message](https://flutter.dev/docs/development/platform-integration/platform-channels) to native code. By default, Flutter only initializes the native messenger after `runApp` was called. If you want to use that messenger before you call `runApp`, you need to call the `ensureInitialized` method or you get the exception you've seen. There's nothing wrong with that approach IMO, it's a valid solution that works. I wouldn't call it a hack.

**Reply by matthewkooshad** [reply]
thanks much for explaining this further; I'm glad to get this working! thank you
